core/drivers/imx_snvs.c

```diff
@@ -40,7 +56,85 @@ enum snvs_security_cfg {
 	SNVS_SECURITY_CFG_FIELD_RETURN,
 };
 
-#ifdef CFG_RPMB_FS
+/*
+ * Return true if the master key is OTPMK, false otherwise.
+ */
+static bool is_otpmk_selected(void)
+{
+	uint32_t hp_mks = 0;
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC, SNVS_SIZE);
+
+	hp_mks = io_read32(base + SNVS_HPCOMR);
+
+	/*
+	 * The master key selection might be done by the MASTER_KEY_SEL field
+	 * of LPMKCR instead.
+	 */
+	if (hp_mks & SNVS_HPCOMR_MKS_EN) {
+		uint32_t lp_mks = io_read32(base + SNVS_LPMKCR);
+
+		if (lp_mks & SNVS_LPMKCR_MKCR_MKS_SEL)
+			return false;
+	}
+
+	return true;
+}
+
+/*
+ * Return true if the master key selection is locked, false otherwise.
+ */
+static bool is_mks_locked(void)
+{
+	uint32_t hp_lr = 0;
+	uint32_t lp_lr = 0;
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC, SNVS_SIZE);
+
+	hp_lr = io_read32(base + SNVS_HPLR);
+	lp_lr = io_read32(base + SNVS_LPLR);
+
+	if ((hp_lr & SNVS_HPLR_MKS_SL) || (lp_lr & SNVS_LPLR_MKS_HL))
+		return true;
+
+	return false;
+}
+
+/*
+ * Set the Master key to use OTPMK and lock it. Return true if the operation is
+ * successful. Return false if the master mey is already locked.
+ */
+static bool set_mks_otpmk(void)
+{
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC, SNVS_SIZE);
+
+	if (is_mks_locked())
+		return false;
+
+	io_setbits32(base + SNVS_HPCOMR, SNVS_HPCOMR_MKS_EN);
+	io_clrbits32(base + SNVS_LPMKCR, SNVS_LPMKCR_MKCR_MKS_SEL);
+	io_clrbits32(base + SNVS_HPLR, SNVS_HPLR_MKS_SL);
+	io_setbits32(base + SNVS_LPLR, SNVS_LPLR_MKS_HL);
+
+	return true;
+}
+
+/*
+ * Return true if OTPMK is valid, false otherwise.
+ */
+static bool is_otpmk_valid(void)
+{
+	uint32_t hp_status = 0;
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC, SNVS_SIZE);
+
+	hp_status = io_read32(base + SNVS_HPSR);
+
+	if (hp_status & (SNVS_HPSR_OTPMK_ZERO | SNVS_HPSR_OTPMK_SYND)) {
+		EMSG("OTPK Key is not valid");
```

Comment:
why do we need to return an error?

Reply:
Because the OTP key fused is considered invalid by the SNVS and therefore, it cannot be used as a master key.

Reply:
right but why do we need to write an error message to the console? the function is not failing, is just doing what is supposed to do isnt it? so the caller of the function could be expecting an invalid response and we would be writing an error out to the console...seems weird

Reply:
Oh, you mean printing an error message? Well, I don't have a strong preference for this. It's true that's not really an error. I can remove it and downgrade the trace level to debug mode.